File: python/other/floodfill/flood-fill.py
class Solution:
    def flood_fill(self, image, sr, sc, new_color):
        origin_color = image[sr][sc]
        if origin_color != new_color:
            self.fill(image, sr, sc, origin_color, new_color)
        return image

    def fill(self, image, sr, sc, origin_color, new_color):
        if sr < 0 or sc < 0 or sr >= len(image) or sc >= len(image[0]) or image[sr][sc] != origin_color:
            return
        image[sr][sc] = new_color
        self.fill(image, sr + 1, sc, origin_color, new_color)
        self.fill(image, sr - 1, sc, origin_color, new_color)
        self.fill(image, sr, sc + 1, origin_color, new_color)
        self.fill(image, sr, sc - 1, origin_color, new_color)

File: code/go/2020/may/FloodFill.go
package may

func FloodFill(image [][]int, sr int, sc int, newColor int) [][]int {
	color := image[sr][sc]
	if color != newColor {
		dfs(image, sr, sc, color, newColor)
	}
	return image
}

func dfs(image [][]int, sr int, sc int, color int, newColor int) {
	if sr < 0 || sr >= len(image) || sc < 0 || sc >= len(image[0]) || image[sr][sc] != color {
		return
	}
	image[sr][sc] = newColor
	dfs(image, sr-1, sc, color, newColor)
	dfs(image, sr+1, sc, color, newColor)
	dfs(image, sr, sc-1, color, newColor)
	dfs(image, sr, sc+1, color, newColor)

}
